_source/_posts/2023-07-28-oidc_workshop.md

```diff
@@ -0,0 +1,505 @@
+  if(!strategy) {
+    return res.sendStatus(404);
+  }
+
+  passport.authenticate(strategy)(req, res, next);
+
+});
+```
+
+For passport to authenticate the user, it will redirect them to their identity provider so that they can log in. After the OpenID server authenticates the user, it redirects them back to the application's `callback` url. 
+
+To recap, a call to the backend's `/openid/start/` endpoint with an org's ID will use passport to redirect the user to their org's OpenID server. The user will prove their identity to their org's OpenID server, and that server will return the authenticated session to the Todo app via passport. Only the app's backend will touch the session information from the OIDC server, so frontend tampering cannot intercept the user's session. 
+
+Once the OpenID server has authenticated the user, it will redirect them to the Todo app backend's `callback` URL. Your next step is to implement the `/openid/callback/${org_id}` endpoint.
+
+## Receive callback after authentication
+
+When the OIDC server redirects to the `callback` URL on the application's backend, Passport will redirect the user back to the app. First, though, the backend should validate that the ID of the request corresponds to a real org in the database, and make the OIDC strategy for that org so Passport can continue to use it. 
+
+Use code like this to implement the `callback` route in `main.ts`: 
+
+```
+app.get('/openid/callback/:id', async (req, res, next) => {
+
+  const org = await orgFromId(req.params.id);
+  if(!org) {
+    return res.sendStatus(404);
+  }
+
+  const strategy = createStrategy(org);
+  if(!strategy) {
+    return res.sendStatus(404);
+  }
+
+  passport.authenticate(strategy, {
+    successRedirect: 'http://localhost:3000/'
+  })(req, res, next);
+
+});
+```
+
+What kinds of problem is this code anticipating? Where will the user be sent if their authentication is successful? 
+
+# Connect to Identity Provider
+
+Now the app is ready for customers to connect their identity providers! This workshop will use an Okta Developer Account as the OIDC provider, to simulate a customer using Okta to connect with your app. 
+
+In a production setting, you would have your customers provide you with information about their identity providers. But in this workshop, you're pretending to be your own customer, so you'll gather the OIDC provider information in this step.
+
+Visit [developer.okta.com](developer.okta.com) and log in to your Developer Account, or sign up if you don't have an account yet. Open the admin console if you're redirected to your user account dashboard. 
+
+## Create App Integration
+
+In the Okta admin console, navigate to "Applications" under the "Applications" heading in the left sidebar. Click the "Create App Integration" button, because your sample app isn't published to the app catalog. (If your app's audience includes a lot of Okta customers, publishing to the Okta catalog can simplify their onboarding process.) 
```

Comment:
```suggestion
In the Okta admin console, navigate to "Applications" under the "Applications" heading in the left sidebar. Click the "Create App Integration" button, because your sample app isn't published to the app catalog. (If your app's audience includes a lot of Okta customers, publishing to the Okta Integration Network can simplify their onboarding process.) 
```